File: VB/WebApplication1/ReportStorageWebExtension1.vb
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Linq
Imports System.ServiceModel
Imports DevExpress.XtraReports.Web.Extensions

Namespace WebApplication1
	Public Class ReportStorageWebExtension1
		Inherits DevExpress.XtraReports.Web.Extensions.ReportStorageWebExtension

		Private ReadOnly reportDirectory As String
		Private Const FileExtension As String = ".repx"
		Public Sub New(ByVal reportDirectory As String)
			If Not Directory.Exists(reportDirectory) Then
				Directory.CreateDirectory(reportDirectory)
			End If
			Me.reportDirectory = reportDirectory
		End Sub
		Public Overrides Function CanSetData(ByVal url As String) As Boolean
			Return Not url.Contains("readonly")
		End Function
		Public Overrides Function IsValidUrl(ByVal url As String) As Boolean
			Return True
		End Function

		Public Overrides Function GetData(ByVal url As String) As Byte()
			Try
                If Directory.EnumerateFiles(reportDirectory).Select(AddressOf Path.GetFileNameWithoutExtension).Contains(url) Then
                    Return File.ReadAllBytes(Path.Combine(reportDirectory, url & FileExtension))
                End If
                Throw New FaultException(New FaultReason(String.Format("Could not find report '{0}'.", url)), New FaultCode("Server"), "GetData")
			Catch e1 As Exception
				Throw New FaultException(New FaultReason(String.Format("Could not find report '{0}'.", url)), New FaultCode("Server"), "GetData")
			End Try
		End Function

		Public Overrides Function GetUrls() As Dictionary(Of String, String)
            Return Directory.GetFiles(reportDirectory, "*" & FileExtension).Select(AddressOf Path.GetFileNameWithoutExtension).ToDictionary(Function(x) x)
        End Function

		Public Overrides Sub SetData(ByVal report As DevExpress.XtraReports.UI.XtraReport, ByVal url As String)
			report.SaveLayoutToXml(Path.Combine(reportDirectory, url & FileExtension))
		End Sub

		Public Overrides Function SetNewData(ByVal report As DevExpress.XtraReports.UI.XtraReport, ByVal defaultUrl As String) As String
			SetData(report, defaultUrl)
			Return defaultUrl
		End Function
	End Class
End Namespace

File: CS/WebApplication1/ReportStorageWebExtension1.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.ServiceModel;
using DevExpress.XtraReports.Web.Extensions;

namespace WebApplication1
{
    public class ReportStorageWebExtension1 : DevExpress.XtraReports.Web.Extensions.ReportStorageWebExtension
    {
        readonly string reportDirectory;
        const string FileExtension = ".repx";
        public ReportStorageWebExtension1(string reportDirectory)
        {
            if (!Directory.Exists(reportDirectory))
            {
                Directory.CreateDirectory(reportDirectory);
            }
            this.reportDirectory = reportDirectory;
        }
        public override bool CanSetData(string url)
        {
            return !url.Contains("readonly");
        }
        public override bool IsValidUrl(string url)
        {
            return true;
        }

        public override byte[] GetData(string url)
        {
            try
            {
                if (Directory.EnumerateFiles(reportDirectory).Select(Path.GetFileNameWithoutExtension).Contains(url))
                {
                    return File.ReadAllBytes(Path.Combine(reportDirectory, url + FileExtension));
                }
                throw new FaultException(new FaultReason(string.Format("Could not find report '{0}'.", url)), new FaultCode("Server"), "GetData");
            }
            catch (Exception)
            {
                throw new FaultException(new FaultReason(string.Format("Could not find report '{0}'.", url)), new FaultCode("Server"), "GetData");
            }
        }

        public override Dictionary<string, string> GetUrls()
        {
            return Directory.GetFiles(reportDirectory, "*" + FileExtension)
                         .Select(Path.GetFileNameWithoutExtension)
                         .ToDictionary<string, string>(x => x);
        }

        public override void SetData(DevExpress.XtraReports.UI.XtraReport report, string url)
        {
            report.SaveLayoutToXml(Path.Combine(reportDirectory, url + FileExtension));
        }

        public override string SetNewData(DevExpress.XtraReports.UI.XtraReport report, string defaultUrl)
        {
            SetData(report, defaultUrl);
            return defaultUrl;
        }
    }
}
